src/util/algorithms.cpp

```diff
@@ -39,3 +41,46 @@ bool goto_functions_algorithm::runOnFunction(
 {
   return true;
 }
+
+void ssa_step_algorithm::run_on_step(symex_target_equationt::SSA_stept &step)
+{
+  // Helper definition
+  typedef goto_trace_stept::typet ssa_type;
+  switch(step.type)
+  {
+  case ssa_type::ASSIGNMENT:
+    run_on_assignment(step);
+    break;
+  case ssa_type::ASSUME:
+    run_on_assume(step);
+    break;
+  case ssa_type::ASSERT:
+    run_on_assert(step);
+    break;
+  case ssa_type::OUTPUT:
+    run_on_output(step);
+    break;
+  case ssa_type::RENUMBER:
+    run_on_renumber(step);
+    break;
+  case ssa_type::SKIP:
+    run_on_skip(step);
+    break;
+  default:
+    log_warning("Invalid type for SSA step");
+  }
+}
+
+void ssa_step_algorithm::iterate(
+  symex_target_equationt::SSA_stepst &eq,
+  bool in_order)
```

Comment:
Do we need this function? The contract is not really clear when `in_order` is false. It could be any order. An algorithm doesn't necessarily have to visit all steps in the list exactly once.

Reply:
> Do we need this function? 

I think so.

> The contract is not really clear when in_order is false

I can try a better naming such as iterate_in_reverse or an enum class DIRECTION { FORWARD, BACKWARD }.

> An algorithm doesn't necessarily have to visit all steps in the list exactly once

Yeah, this is mostly a helper function though, we also have the `run_on_step` in case
that the algorithm needs to do a custom flow.

Reply:
> we also have the run_on_step in case that the algorithm needs to do a custom flow.

Right, that's why this function seems pretty useless to me. Shortening a long `switch` could be nice, but a 2-line `for` loop? Not sure a function call to a base class makes that easier to read.

Reply:
In this PR, this method seems unnecessary since we always call the method `iterate` with the `in_order` set to false. So, I agree with @fbrausse here.

@rafaelsamenezes: Is there some other class in our codebase that needs to use this method by setting `in_order` to true?

Reply:
It will be used here at PR https://github.com/esbmc/esbmc/pull/827

Reply:
A function that branches off to entirely different behavious based on a command line parameter is not well-defined. There is no need to encapsulate a for-loop that just delegates to another function.

Reply:
I removed the iterate method